ydb/library/actors/wilson/wilson_uploader.cpp

```diff
@@ -90,25 +156,56 @@ namespace NWilson {
                 if (SpansSize >= 100'000'000) {
                     LOG_ERROR_S(*TlsActivationContext, WILSON_SERVICE_ID, "dropped span due to overflow");
                 } else {
-                    const TMonotonic expirationTimestamp = TActivationContext::Monotonic() + MaxSpanTimeInQueue;
+                    const TMonotonic now = TActivationContext::Monotonic();
+                    const TMonotonic expirationTimestamp = now + MaxSpanTimeInQueue;
                     auto& span = ev->Get()->Span;
                     const ui32 size = span.ByteSizeLong();
-                    Spans.push_back(TSpanQueueItem{expirationTimestamp, std::move(span), size});
+                    if (size > MaxBytesInBatch) {
+                        ALOG_ERROR(WILSON_SERVICE_ID, "dropped span of size " << size << ", which exceeds max batch size " << MaxBytesInBatch);
+                        return;
+                    }
+                    TSpan spanItem {
+                        .ExpirationTimestamp = expirationTimestamp,
+                        .Span = std::move(span),
+                        .Size = size,
+                    };
                     SpansSize += size;
+                    if (CurrentBatch.IsEmpty()) {
+                        ScheduleBatchCompletion(now);
+                    }
+                    if (CurrentBatch.Add(spanItem)) {
+                        return;
+                    }
+                    CompleteCurrentBatch();
                     TryMakeProgress();
+                    Y_ABORT_UNLESS(CurrentBatch.Add(spanItem), "failed to add span to empty batch");
+                    ScheduleBatchCompletion(now);
                 }
             }
 
+            void ScheduleBatchCompletion(TMonotonic now) {
+                TMonotonic completionTime = now + TDuration::MilliSeconds(MaxBatchAccumulationMilliseconds);
+                TActivationContext::Schedule(completionTime,
```

Comment:
This may lead to a big amount of scheduled messages when the batches are issued rapidly by the size. It's better to keep inflight of 1 and reschedule message when it's not the time to finalize the batch.

Reply:
Well, it wouldn't lead to more, than spans/batch_size messages, but yep, that's a great idea, fixed